### PR TITLE
[feature](json-function):Add json_merge_preserve

### LINKED ISSUE #21268: [Enhancement] Improve Json function for doris
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Now We have some _JSONB functions_ and  _JSON functions_ are not implement , specific semantics for each function below here can reference mysql

**NOTE** : _JSONB Functions_ , which has same function name with _JSON functions_ but with different implement which can be find in `be/src/vec/functions/function_jsonb.cpp ` . eg. json_parse alias jsonb_parse

### about json_path define:
https://modern-sql.com/blog/2017-06/whats-new-in-sql-2016#json-path

#### we have some JSONB functions which are not implement but JSON has already implement. below here
- [ ]  json_contains  @weijiew 
- [ ]  json_contains_path @weijiew 
- [ ]  json_quote json_unquote 
- [ ]  json_valid @hammerZh-Z 
- [ ]  json_array @yifei02 
- [ ]  json_object

#### below here we have both JSONB and JSON functions not implement .

- [ ]  ->>  ([->>](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#operator_json-inline-path))
- [ ] JSON_DEPTH ([JSON_DEPTH()](https://dev.mysql.com/doc/refman/8.0/en/json-attribute-functions.html#function_json-depth)) @niebayes 
- [ ] JSON_LENGTH ([JSON_LENGTH()](https://dev.mysql.com/doc/refman/8.0/en/json-attribute-functions.html#function_json-length)) @Xuxiaotuan 
- [ ] JSON_KEYS ([JSON_KEYS()](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#function_json-keys)) @caicancai 
- [ ] JSON_SEARCH ([JSON_SEARCH()](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#function_json-search)) 
- [ ] MERGE_OF ([MEMBER OF()](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#operator_member-of))

- [ ] JSON_INSERT ([JSON_INSERT()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-insert)).  @JasonWen1 
- [ ] JSON_REPLACE ([JSON_REPLACE()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-replace)) @cjj2010 
- [ ] JSON_SET ([JSON_SET()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-set))

- [ ] JSON_OVERLAPS ([JSON_OVERLAPS()](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#function_json-overlaps))
- [ ] JSON_MERGE_PATCH ([JSON_MERGE_PATCH()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-merge-patch))
- [ ] JSON_MERGE_PRESERVE ([JSON_MERGE_PRESERVE()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-merge-preserve)) @taptao 
- [ ] JSON_REMOVE ([JSON_REMOVE()](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-remove))
- [ ] JSON_TABLE ([JSON_TABLE()](https://dev.mysql.com/doc/refman/8.0/en/json-table-functions.html#function_json-table)) @lovlyhao 
- [ ] JSON_STORAGE_SIZE ([JSON_STORAGE_SIZE()](https://dev.mysql.com/doc/refman/8.0/en/json-utility-functions.html#function_json-storage-size))
- [ ] JSON_STORAGE_FREE ([JSON_STORAGE_FREE()](https://dev.mysql.com/doc/refman/8.0/en/json-utility-functions.html#function_json-storage-free))

U can pick this according comment below:
"I pick "-> ->> " " and this task could assign to u for one week, if no pr or progress during one week I would re-assign others

### Solution

U can pick this according comment below:
"I pick "-> ->> " " and this task could assign to u for one week, if no pr or progress during one week I would re-assign others 

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by taptao**
## Proposed changes

Issue Number: close https://github.com/apache/doris/issues/21268

## Further comments
Add json_merge_preserve function.

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by taptao** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.12 seconds
 stream load tsv:          512 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.5 seconds inserted 10000000 Rows, about 338K ops/s
 storage size: 17162384434 Bytes

File: regression-test/suites/query_p0/sql_functions/json_functions/test_json_function.groovy [hunk 1]
```diff
@@ -83,4 +83,14 @@ suite("test_json_function") {
     qt_sql "SELECT JSON_CONTAINS('{\"name\": \"John\", \"age\": 30, \"projects\": [{\"name\": \"Project A\", \"year\": 2020}, {\"name\": \"Project B\", \"year\": 2021}]}', '{\"projects\": [{\"name\": \"Project A\"}]}', '\$.');"
     qt_sql "SELECT JSON_CONTAINS('{\"name\": \"John\", \"age\": 30, \"address\": {\"city\": \"New York\", \"country\": \"USA\"}}', '{\"address\": {\"city\": \"New York\"}}', '\$.');"
 
+    qt_sql "SELECT JSON_MERGE_PRESERVE('[]','[]');"
```
Comment (amorynan) [inline regression-test/suites/query_p0/sql_functions/json_functions/test_json_function.groovy:86]: make same test json in table

File: be/src/vec/functions/function_json.cpp [hunk 2]
```diff
@@ -936,6 +937,186 @@ class FunctionJsonValid : public IFunction {
...
+                rapidjson::Value* node =
+                        static_cast<rapidjson::Value*>(allocator.Malloc(sizeof(rapidjson::Value)));
+                node->CopyFrom(*json, allocator);
+                result->PushBack(*node, allocator);
+            }
+        }
+        return result;
+    }
+
+    rapidjson::Value* merge_obejcts(std::vector<rapidjson::Value*> objs,
```
Comment (amorynan) [inline be/src/vec/functions/function_json.cpp:981]: write some step or description of ur function

File: be/src/vec/functions/function_json.cpp [hunk 3]
```diff
@@ -936,6 +937,186 @@ class FunctionJsonValid : public IFunction {
...
+                                  rapidjson::Document::AllocatorType& allocator) {
+        rapidjson::Value* result =
+                static_cast<rapidjson::Value*>(allocator.Malloc(sizeof(rapidjson::Value)));
+
+        if (jsons.empty()) {
+            return result;
+        }
+
+        if (jsons.size() == 1) {
+            return jsons.at(0);
```
Comment (amorynan) [inline be/src/vec/functions/function_json.cpp:1017]: just jsons[0] is ok

File: regression-test/suites/query_p0/sql_functions/json_functions/test_json_function.groovy (hunk 1 above)
Comment (taptao) [reply regression-test/suites/query_p0/sql_functions/json_functions/test_json_function.groovy:86]: done

File: be/src/vec/functions/function_json.cpp (hunk 2 above)
Comment (taptao) [reply be/src/vec/functions/function_json.cpp:985]: done

File: be/src/vec/functions/function_json.cpp (hunk 3 above)
Comment (taptao) [reply be/src/vec/functions/function_json.cpp]: done

**Comment by taptao** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.35 seconds
 stream load tsv:          542 seconds loaded 74807831229 Bytes, about 131 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          98.0 seconds inserted 10000000 Rows, about 102K ops/s
 storage size: 17162244425 Bytes

**Pull request closed without merging**